Memory corruption(? well, probably not; random name though) -> inscrutable error on plain unpack

This is on a clean new user:
```
c-i-t@tarta:~$ wget https://github.com/ryankurte/cargo-binstall/releases/latest/download/cargo-binstall-x86_64-unknown-linux-musl.tgz
...
2022-08-03 17:23:47 (2.33 MB/s) - ‘cargo-binstall-x86_64-unknown-linux-musl.tgz’ saved [2492469/2492469]

c-i-t@tarta:~$ tar -tvaf cargo-binstall-x86_64-unknown-linux-musl.tgz
-rw-r--r-- runner/docker 35149 2022-07-24 10:48 LICENSE.txt
-rw-r--r-- runner/docker 10668 2022-07-24 10:48 README.md
-rwxr-xr-x runner/docker 5264960 2022-07-24 10:48 cargo-binstall
c-i-t@tarta:~$ tar -xvaf cargo-binstall-x86_64-unknown-linux-musl.tgz  cargo-binstall
cargo-binstall
c-i-t@tarta:~$ ./cargo-binstall
15:24:11 [ERROR] Fatal error:
binstall::io (https://docs.rs/cargo-binstall/0.11.1/cargo_binstall/enum.BinstallError.html#variant.Io)

  × No such file or directory (os error 2) at path "/home/c-i-t/.cargo/bin/cargo-binstalleMdRRS"
```

It looks like memory corruption but probably isn't, considering, so it's probably some random name? Either way, not sure what's supposed to happen here, but it's not that.

[`strace -f`](https://github.com/ryankurte/cargo-binstall/files/9252909/ss.gz); --log-level=debug:
```
c-i-t@tarta:~$ ./cargo-binstall --log-level=debug
15:29:28 [DEBUG] (25) cargo_binstall::helpers: using (/home/c-i-t/.cargo) as cargo home
15:29:28 [DEBUG] (25) cargo_binstall: Using install path: /home/c-i-t/.cargo/bin
15:29:28 [DEBUG] (1) cargo_binstall: run time: 3.040425ms
15:29:28 [ERROR] Fatal error:
binstall::io (https://docs.rs/cargo-binstall/0.11.1/cargo_binstall/enum.BinstallError.html#variant.Io)

  × No such file or directory (os error 2) at path "/home/c-i-t/.cargo/bin/cargo-binstallIwSQUG"
```

I created that directory, then I got
```
c-i-t@tarta:~$ ./cargo-binstall
Do you wish to continue? yes/[no]
? yes
15:52:00 [ERROR] Fatal error:

  × missing field `v1`
```

This is very good because (a) it doesn't even say where the error is, (b) cargo-binstall literally O_CREAT|O_EXCLs .crates.toml, and (c) `cargo install` creates it on first use in a similar fashion, equally empty, so this is a valid configuration.

--- a/src/main.rs
+++ b/src/main.rs
@@ -1,5 +1,6 @@
 use std::{
     ffi::OsString,
+    fs,
     mem::take,
     path::{Path, PathBuf},
     process::{ExitCode, Termination},
@@ -295,6 +296,7 @@ async fn entry(jobserver_client: LazyJobserverClient) -> Result<()> {
         error!("No viable install path found of specified, try `--install-path`");
         miette!("No install path found or specified")
     })?;
+    fs::create_dir_all(&install_path).map_err(BinstallError::Io)?;
     debug!("Using install path: {}", install_path.display());
 
     // Create a temporary directory for downloads etc.
@@ -393,6 +395,10 @@ async fn entry(jobserver_client: LazyJobserverClient) -> Result<()> {
 
     block_in_place(|| {
         if !custom_install_path {
+            // If using standardised install path,
+            // then create_dir_all(&install_path) would also
+            // create .cargo.
+
             debug!("Writing .crates.toml");
             metafiles::v1::CratesToml::append(metadata_vec.iter())?;
 

--- a/src/metafiles/v1.rs
+++ b/src/metafiles/v1.rs
@@ -74,7 +74,11 @@ impl CratesToml {
         Iter: IntoIterator<Item = &'a MetaData>,
     {
         let mut file = FileLock::new_exclusive(create_if_not_exist(path.as_ref())?)?;
-        let mut c1 = Self::load_from_reader(&mut *file)?;
+        let mut c1 = if file.metadata()?.len() != 0 {
+            Self::load_from_reader(&mut *file)?
+        } else {
+            Self::default()
+        };
 
         for metadata in iter {
             c1.insert(&CrateVersionSource::from(metadata), metadata.bins.clone());
@@ -107,4 +111,33 @@ pub enum CratesTomlParseError {
 
     #[error(transparent)]
     CvsParse(#[from] super::CvsParseError),
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{super::binstall_v1, *};
+    use crate::target::TARGET;
+
+    use semver::Version;
+    use tempfile::TempDir;
+
+    #[test]
+    fn test_empty() {
+        let tempdir = TempDir::new().unwrap();
+        let path = tempdir.path().join("crates-v1.toml");
+
+        CratesToml::append_to_path(
+            &path,
+            &[MetaData {
+                name: "cargo-binstall".into(),
+                version_req: "*".into(),
+                current_version: Version::new(0, 11, 1),
+                source: binstall_v1::Source::cratesio_registry(),
+                target: TARGET.into(),
+                bins: vec!["cargo-binstall".into()],
+                other: Default::default(),
+            }],
+        )
+        .unwrap();
+    }
 }

